ICU-22954 USet C++ iterator return std::u16string; header-only LocalPointer

markusicu opened the pull request:
- ICU-22954 Build error in uset.h when `U_SHOW_CPLUSPLUS_HEADER_API` is true but `U_SHOW_CPLUSPLUS_API` is false
    - USet C++ iterator return std::u16string instead of a not-header-only icu::UnicodeString
    - Change USetElements to draft ICU 77 because of this change
    - Make the intltest C++ test framework work without U_SHOW_CPLUSPLUS_API by
        - switching from UnicodeString arguments to std::u16string_view
        - decoupling IcuTestErrorCode from the DLL-exported class ErrorCode
    - Test header-only USet C++ iterators in an intltest file with U_SHOW_CPLUSPLUS_API=0
- ICU-22980 make localpointer.h header-only compatible, and parts of char16ptr.h
    - Add U_ICU_NAMESPACE_OR_INTERNAL = icu or icu::internal
    - Change localpointer.h to be header-only and work with just U_SHOW_CPLUSPLUS_HEADER_API
    - (Const)Char16Ptr need to remain DLL-exported because they are baked into DLL-exported APIs in UnicodeString member function signatures

Best to review one commit at a time.

#### Checklist
- [x] Required: Issue filed: ICU-22954
- [x] Required: The PR title must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

ALLOW_MANY_COMMITS=true

roubert commented on icu4c/source/test/intltest/usettest.cpp:
```diff
@@ -4283,35 +4279,7 @@ void UnicodeSetTest::TestCodePointIterator() {
-    assertEquals(WHERE, U'🚴', *iter3);
-    assertTrue(WHERE, iter2 == limit);
-}
+// TestUSetCodePointIterator moved to headeronlytest.cpp
```
I doubt that future readers will find such comments helpful, most of them will be unlikely to know that these test cases ever where here in the first place.

markusicu replied:
ok done

roubert commented on icu4c/source/test/intltest/headeronlytest.cpp:
```diff
@@ -0,0 +1,213 @@
+    LocalUSetPointer uset(uset_openPattern(u"[abcçカ🚴{}{abc}{de}]", -1, errorCode));
+    std::u16string result;
+    for (auto s : USetStrings(uset.getAlias())) {
+        // UnicodeString us(s);
```
This should be `std::u16string` now, right? (Also the same further down.)

markusicu replied:
no: `auto s` is now a std::u16string, so in order to make the .toUTF8String() two lines below work, we need to convert it to a UnicodeString

previously, in usettest.cpp, this line wasn't there, because the s was already a UnicodeString

roubert replied:
But converting to `UnicodeString` is impossible here, there's a `#define U_SHOW_CPLUSPLUS_API 0` at the beginning of the file, `UnicodeString` doesn't exist here.

markusicu replied:
This is commented-out sample code. Someone who wants to run that, like I did to convince myself that it still worked, has to turn U_SHOW_CPLUSPLUS_API back on. That's what I did for updating the sample code snippets in the API docs.

roubert replied:
> Someone who wants to run that, like I did to convince myself that it still worked, has to turn U_SHOW_CPLUSPLUS_API back on.

OK … that's something that I did not manage to guess on my own. What do the other reviewers say here: Was this obvious to you?

markusicu replied:
added comments about sample code for pasting

roubert commented on icu4c/source/common/unicode/uset.h:
```diff
@@ -1756,26 +1759,28 @@ class USetElementIterator {
+    std::u16string operator*() const {
         if (c >= 0) {
-            return UnicodeString(c);
+            return c <= 0xffff ?
```
Why not use the `U_IS_BMP()` macro here too, like in the test code?

markusicu replied:
i don't think it really matters one way or another; slight preference not to add an `include unicode/utf.h` to uset.h

roubert replied:
It's just odd to me to do the same thing in two different ways in the same PR for no obvious reason.

markusicu replied:
Well, so I gave is a mild reason.
Either way the code only makes sense to someone who understands code points and UTF-16, and if they do, then both versions should be equally fine.

roubert commented on icu4c/source/common/unicode/uset.h:
```diff
@@ -1756,26 +1759,28 @@ class USetElementIterator {
             const UChar *uchars = uset_getString(uset, index - rangeCount, &length);
             // assert uchars != nullptr;
-            return UnicodeString(uchars, length);
+            return {ConstChar16Ptr(uchars), static_cast<uint32_t>(length)};
```
Shouldn't that be `std::u16string::size_type` instead of `uint32_t`?

markusicu replied:
i prefer just casting to the unsigned version of the type and letting the constructor do the rest

if anything i would cast to size_t -- i find the type-specific something::size_type super clunky and unnecessary

eggrobin replied:
Since we need to have a cast anyway, I would prefer casting to the right thing directly. I agree that size_t is more readable though (and it follows from https://eel.is/c++draft/string.classes#general-1, https://eel.is/c++draft/string.classes#basic.string.general-3, and https://eel.is/c++draft/default.allocator#general-1 that `std::u16string::size_type` is `std::size_t`).

markusicu replied:
fine, will do

markusicu commented:
@richgillam sort-of-approved the API change proposal. I don't expect any controversy.

It would be lovely to get this PR reviewed (I suggest one commit at a time) and approved so that I can merge it soon after the upcoming ICU-TC meeting.

markusicu commented:
Oops, this window hadn't refreshed before I sent my previous comment -- now it has. Will respond to feedback.

markusicu commented on icu4c/source/test/intltest/headeronlytest.cpp:
```diff
@@ -0,0 +1,213 @@
+
+// Test header-only ICU C++ APIs. Do not use other ICU C++ APIs.
+#define U_SHOW_CPLUSPLUS_API 0
+#define U_SHOW_CPLUSPLUS_HEADER_API 1
```
I prefer these two to be explicit.

markusicu replied:
commenting it out now

markusicu commented on icu4c/source/test/intltest/headeronlytest.cpp:
```diff
@@ -0,0 +1,213 @@
+#include "unicode/utf16.h"
+#include "intltest.h"
+
+using namespace std::string_view_literals;
```
It might, in some cases, but while trying to make intltest work, I had a bit of back-and-forth with adding assertEquals() overloads and disambiguating call sites.

In my view, it's safe and harmless, and at most a bit overly specific.

eggrobin replied:
We really should just templatize assertEquals and make it work with heterogeneous types.
Not a blocker for this PR whose scope has already kept quite far obviously, but feel free to assign something to me to do that.

roubert replied:
Improving `assertEquals()` seems like a good idea but that's not something that's needed for this PR, just using `u""` without involving any string view literals works just fine and is simpler.

markusicu replied:
I am removing the `sv` prefix now, and verified that it does work.
However, I don't understand what's *wrong* about using string_view literals.

roubert replied:
> I don't understand what's _wrong_ about using string_view literals.

It's an unnecessary complication that distracts from the real purpose of the code. You can, of course, always add additional code that does unnecessary but harmless things, but you normally don't because it makes your code harder to read.

markusicu commented on icu4c/source/test/intltest/headeronlytest.cpp:
```diff
@@ -0,0 +1,213 @@
+
+class HeaderOnlyTest : public IntlTest {
+public:
+    HeaderOnlyTest() {}
```
ok, done (not sure it makes any difference)

eggrobin replied:
There is a big difference between a defaulted constructor (explicitly or implicitly) and a user-provided constructor. The former is [_trivial_](https://eel.is/c++draft/class.default.ctor#def:constructor,default,trivial), which matters for many language rules.

markusicu replied:
Interesting -- but subclasses of IntlTest don't fit several of the criteria for "trivial", so it probably makes no real difference.

eggrobin replied:
Indeed, it’s just that `{}` raises questions for a C++ reviewer, whereas `= default` is, well, the default :-)

markusicu commented on icu4c/source/test/intltest/headeronlytest.cpp:
```diff
@@ -0,0 +1,213 @@
+    void TestUSetElementIterator();
+};
+
+extern IntlTest *createHeaderOnlyTest() {
```
it makes it explicit that it's called from another file

markusicu commented on icu4c/source/test/intltest/Makefile.in:
```diff
@@ -75,7 +75,7 @@ numbertest_parse.o numbertest_doubleconversion.o numbertest_skeletons.o \
 formattedvaluetest.o formatted_string_builder_test.o numbertest_permutation.o \
 units_data_test.o units_router_test.o units_test.o displayoptions_test.o \
-numbertest_simple.o uchar_type_build_test.o
+numbertest_simple.o uchar_type_build_test.o headeronlytest.o
```
i gave it a generic name so that we can collect multiple header-only tests here

roubert replied:
That doesn't seem like a good idea to me, it'll be easier to navigate the tests if each set of tests is in their own file, with this one for the USet tests.

markusicu replied:
ok done

markusicu commented:
> Will respond to feedback.

done

markusicu commented:
feedback addressed, PTAL

roubert commented on icu4c/source/test/intltest/headeronlytest.cpp:
```diff
@@ -0,0 +1,213 @@
+// created: 2024dec11 Markus W. Scherer
+
+#include <string>
+#include <string_view>
```
This is now no longer used.

markusicu replied:
gone

eggrobin commented on icu4c/source/test/intltest/usetheaderonlytest.cpp:
```diff
@@ -0,0 +1,219 @@
+    std::u16string result;
+    for (auto el : USetElements(uset.getAlias())) {
+        // Commented-out sample code for pasting into the API docs.
+        // Needs U_SHOW_CPLUSPLUS_API=1 for UnicodeString.
```
I am a bit confused.

If we are giving sample code for an API that does not require the C++ API, shouldn’t we use a sample that doesn’t use it ?

And if we are giving sample code that does require this API, shouldn’t we test it in the other test where we do require the C++ API ?

markusicu replied:
1. I am a bit confused why we are having an extended discussion about pre-existing, commented-out sample code. I am just moving it from one file to another, and adding one more commented-out line to make it work when not commented out.
2. The sample code is using UnicodeString for conversion to UTF-8. I concede that this does not work for someone using U_SHOW_CPLUSPLUS_API=0, but it makes for nice, succinct sample code.
3. I don't know how to output a UTF-16 string in portable C++ when not using ICU. Let me know if you have suggestions.
4. I suppose I could change this sample code to calling ICU C (ucnv.h) APIs, but then there would be more code dealing with conversion than showing how to use the iteration API.
5. I suppose I could just omit the print statement and write a comment for what the UTF-16 string will contain in each iteration.
6. Can we please deal with the commented-out sample code in a follow-up pull request?

eggrobin replied:
Re 6, sure.

markusicu replied:
I just took a look:
- https://en.cppreference.com/w/cpp/locale/codecvt_utf8_utf16 -- zombie: added in C++11, deprecated in C++17, removed in C++26
- https://en.cppreference.com/w/cpp/locale/codecvt -- zombie:
  - std::codecvt<char16_t, char, std::mbstate_t> was added in C++11 but deprecated in C++20
  - std::codecvt<char16_t, char8_t, std::mbstate_t> was apparently born deprecated in C++20

markusicu replied:
PS: I guess the most obvious thing that *should just work*™ would be `cout << s` where s is a std::u16string -- but I am pretty sure that that does not work.

markusicu replied:
Fixed in
- https://github.com/unicode-org/icu/pull/3316

roubert replied:
> 1. I am a bit confused why we are having an extended discussion about pre-existing, commented-out sample code.

Because the context of that code was changed.

markusicu commented:
FYI: The ICU-TC approved the API change proposal.

markusicu commented:
Thanks Robin!

@roubert I have addressed your feedback. I don't want to fiddle further with the sample code in this PR. If I don't hear back from you during Japan business hours, then I plan to dismiss your change request and merge.

Except, I will first try to squash the feedback commits into the regular ones, so that this is down to a small number of somewhat meaningful commits.

@richgillam please speak up if you disagree.

richgillam commented:
> @richgillam please speak up if you disagree.

No, I think I'm happy...
